compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java

```diff
@@ -1193,10 +1196,16 @@ void addLiteralValue(DiagnosticPos pos, Set<Whitespace> ws, int typeTag, Object
     }
 
     void addLiteralValue(DiagnosticPos pos, Set<Whitespace> ws, int typeTag, Object value, String originalValue) {
-        BLangLiteral litExpr = (BLangLiteral) TreeBuilder.createLiteralExpression();
+        BLangLiteral litExpr;
+        // If numeric literal create a numeric literal expression; otherwise create a literal expression
+        if (typeTag < TypeTags.DECIMAL) {
+            litExpr = (BLangNumericLiteral) TreeBuilder.createNumericLiteralExpression();
+        } else {
+            litExpr = (BLangLiteral) TreeBuilder.createLiteralExpression();
+        }
         litExpr.addWS(ws);
         litExpr.pos = pos;
-        litExpr.typeTag = typeTag;
+        litExpr.type = new BType(typeTag, null);
```

Comment:
type should be taken from the symbolTable.

Applicable for all places

Reply:
Since we are at the PackageBuilder level we can't get the type from the SymbolTable for this case. As we are removing the `typeTag` field from the `BLangLiteral` to hold the `type/type tag` information, we have to use the `type` field. However, at this level the type is null. 
To tackle this we require to create a `new BType()` here, which will anyway be modified always at the `TypeChecker` phase.

Reply:
> Since we are at the PackageBuilder level we can't get the type from the SymbolTable for this case.

Why not? :) 
The catch is, value types are singletons, but here we are creating new instances of same type.

It seems the cleanest approach is to have separate kinds of nodes for different literal types. At the moment we are losing this information at packageBuilder, and trying to fix it by the 'type' has caused these complications.. 
But lets do that as an improvement separately.

Reply:
+1; As you suggested the best way to handle this to have separate node kinds for different literal types. Type's tag and Node kind need to be separated out and type related checks/changes should happen only after type checker phase.